scalanativelib/src/mill/scalanativelib/ScalaNativeModule.scala

```diff
@@ -0,0 +1,275 @@
+import mill.{PathRef, T}
+import mill.util.Loose
+import sbt.testing.{AnnotatedFingerprint, Framework, SubclassFingerprint}
+import sbt.testing.Fingerprint
+import testinterface.ScalaNativeFramework
+import mill.util.Loose.Agg
+
+
+trait ScalaNativeModule extends scalalib.ScalaModule { outer =>
+  def scalaNativeVersion: T[String]
+  override def platformSuffix = s"_native${scalaNativeBinaryVersion()}"
+  override def artifactSuffix: T[String] = s"${platformSuffix()}_${artifactScalaVersion()}"
+
+  trait Tests extends TestScalaNativeModule {
+    override def scalaWorker = outer.scalaWorker
+    override def scalaVersion = outer.scalaVersion()
+    override def scalaNativeVersion = outer.scalaNativeVersion()
+    override def moduleDeps = Seq(outer)
+  }
+
+  def scalaNativeBinaryVersion = T{ scalaNativeVersion().split('.').take(2).mkString(".") }
+
+  def bridge = T.task{ ScalaNativeBridge.scalaNativeBridge().bridge(bridgeFullClassPath()) }
+
+  def scalaNativeBridgeClasspath = T {
+    val snBridgeKey = "MILL_SCALANATIVE_BRIDGE_" + scalaNativeBinaryVersion().replace('.', '_').replace('-', '_')
+    val snBridgePath = sys.props(snBridgeKey)
+    if (snBridgePath != null) Result.Success(
+      Agg(PathRef(Path(snBridgePath), quick = true))
+    ) else Lib.resolveDependencies(
+      Seq(Cache.ivy2Local, MavenRepository("https://repo1.maven.org/maven2")),
+      Lib.depToDependency(_, "2.12.4", ""),
+      Seq(
+        ivy"com.lihaoyi::mill-scalanativelib-scalanativebridges-${scalaNativeBinaryVersion()}:${sys.props("MILL_VERSION")}"
+      )
+    ).map(_.filter(_.path.toString.contains("mill-scalanativelib-scalanativebridges")))
+  }
+
+  def toolsIvyDeps = T{
+    Seq(
+      ivy"org.scala-native:tools_2.12:${scalaNativeVersion()}",
+      ivy"org.scala-native:util_2.12:${scalaNativeVersion()}",
+      ivy"org.scala-native:nir_2.12:${scalaNativeVersion()}"
+    )
+  }
+
+  override def transitiveIvyDeps: T[Agg[Dep]] = T{
+    ivyDeps() ++ nativeIvyDeps() ++ Task.traverse(moduleDeps)(_.transitiveIvyDeps)().flatten
+  }
+
+  def nativeLibIvy = T{ ivy"org.scala-native::nativelib::${scalaNativeVersion()}" }
+
+  def nativeIvyDeps = T{
+    Seq(nativeLibIvy()) ++
+    Seq(
+      ivy"org.scala-native::javalib::${scalaNativeVersion()}",
+      ivy"org.scala-native::auxlib::${scalaNativeVersion()}",
+      ivy"org.scala-native::scalalib::${scalaNativeVersion()}"
+    )
+  }
+
+  def bridgeFullClassPath = T {
+    Lib.resolveDependencies(
+      Seq(Cache.ivy2Local, MavenRepository("https://repo1.maven.org/maven2")),
+      Lib.depToDependency(_, scalaVersion(), platformSuffix()),
+      toolsIvyDeps()
+    ).map(t => (scalaNativeBridgeClasspath().toSeq ++ t.toSeq).map(_.path))
+  }
+
+  override def scalacPluginIvyDeps = super.scalacPluginIvyDeps() ++
+    Agg(ivy"org.scala-native:nscplugin_${scalaVersion()}:${scalaNativeVersion()}")
+
+  def releaseMode = T { false }
```

Comment:
We reserve right to introduce more modes, that's why this setting is a string in the build api and sbt plugin. I'd recommend to keep it a string here as well to avoid future problems if more modes are introduced.

Reply:
OK. I'd been meaning to look at this. Will change.

Reply:
To me it is better to have a sealed trait instead of a string. So if more modes are introduced there are no problems in the future. It is how I did with scala js `ModuleKind`s miming the scalajs api.